reduce boost size

spahrenk opened the pull request:
## Description

- use boost bcp utility to eliminate unused files
- add update-boost.py script for updating/reducing boost

## Author Progress Checklist:

- [x] Open draft pull request
    - [x] Make title clearly understandable in a standalone change log context
    - [x] Assign yourself the issue
    - [x] Add at least one label (enhancement, bug, or maintenance)
    - [x] Link the issue(s) addressed by this PR (under "Development" in the sidebar menu) 
- [x] Make code changes (if you haven't already)
- [x] Self-review of code
    - [x] My code follows the style guidelines of this project
    - [x] I have added comments to my code, particularly in hard-to-understand areas
    - [x] I have only committed the necessary changes for this fix or feature
    - [x] I have made corresponding changes to the documentation
    - [x] My changes generate no new warnings
    - [x] I have ensured that my fix is effective or that my feature works as intended by:
        - [x] exercising the code changes in the test framework, and/or
        - [x] manually verifying the changes (as explained in the the pull request description above)
    - [x] My changes pass all local tests
    - [x] My changes successfully passes CI checks
    - [ ] Add any unit test for proof and documentation.
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
- [ ] Move pull request out of draft mode and assign reviewers
- [ ] Iterate with reviewers until all changes are approved
    - [ ] Make changes in response to reviewer comments
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
    - [ ] Re-request review in GitHub

## Reviewer Checklist:

 - [ ] Read the pull request description
 - [ ] Perform a code review on GitHub
 - [ ] Confirm all CI checks pass and there are no build warnings
 - [ ] Pull, build, and run automated tests locally
 - [ ] Perform manual tests of the fix or feature locally
 - [ ] Add any review comments, if applicable
 - [ ] Submit review in GitHub as either
     - [ ] Request changes, or
     - [ ] Approve
 - [ ] Iterate with author until all changes are approved

spahrenk committed:
reduce boost size

Ran boost bcp utility to eliminate unused files.

spahrenk committed:
add update-boost script and apply

Commit:
modified mgl file define.h to avoid _snprintf error wwith msvc

spahrenk committed:
potential fix for gcc compiler warning with boost

nealkruis commented on src/libkiva/Geometry.hpp:
```diff
@@ -20,6 +20,7 @@
 #pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
 #pragma GCC diagnostic ignored "-Wuninitialized"
 #pragma GCC diagnostic ignored "-Wstrict-overflow"
+#pragma GCC system_header
```
This is maybe too broad of a solution. What are the specific warnings that came up? Is it specific to a recent version of GCC? Are the boost developers aware of the warning?

spahrenk replied:
Used boost macros to disable quadmath and avoid warnings.

nealkruis commented on scripts/update-boost.py:
```diff
@@ -0,0 +1,85 @@
+source_boost_path = docs_path + "/Development/boost_1_82_0"
+
+# path to repo
+repo_path = docs_path + "/GitHub/kiva"
```
This seems specific to your machine's setup. It should be generalized to work on any machine and any platform.

spahrenk replied:
Changed script to enable command-line use and cross-platform compatibility.

spahrenk committed:
disable boost quadmath

spahrenk committed:
Revert to boost1_81_0; make update script cross-platform

spahrenk committed:
Update update-boost.py

spahrenk commented:
Modified script for cross-platform compatibility. Reverted to boost-1.81.0 to avoid compile warning in later version.

spahrenk commented:
Reverted to boost-1.81.0 to avoid compiler warning in 1.82.0

spahrenk committed:
Address Windows issues

spahrenk committed:
update to boost1.83, fix script, modify macro defs

Issue with Windows using earlier boost versions. Adjusted quadlib macro defs in Geometry.hpp to apply to all builds.

spahrenk committed:
correct boost typo; improve script

tanaya-mankad commented on scripts/update-boost.py:
```diff
@@ -0,0 +1,195 @@
+	for repo_src_path, dirs, files in os.walk(repo_src_path, topdown = True):
+		for filename in files:
+			full_filename = os.path.join(repo_src_path, filename)
+			file_list = file_list + " \"" + full_filename + "\""
```
I don't think the string `file_list` is subsequently used.

tanaya-mankad commented on scripts/update-boost.py:
```diff
@@ -0,0 +1,195 @@
+		os.mkdir(dest_repo_boost_path )
+	
+	# adjust path if Windows
+	if platform.system() == "Windows":
```
I'm pretty sure that this is not an intended use of platform.system(). :) Have a look at Python's pathlib library for platform-independent, object-oriented path names. But I think the os.path module also should be able to deal with this natively.

spahrenk committed:
remove extraneous imports

tanaya-mankad commented on scripts/update-boost.py:
```diff
@@ -116,22 +114,13 @@ def updateBoost(bcp_command, source_boost_path, repo_path, prev_repo_boost_folde
 	# run bcp utility
 	bcp_list = [bcp_command, "--scan"]
 	bcp_list.append("--boost=" + source_boost_path)
 	file_list = ""
```
This is probably an unused variable now.

tanaya-mankad commented on scripts/update-boost.py:
```diff
@@ -0,0 +1,184 @@
+	print(result.stdout)
+	
+	# copy CMakeLists.txt from stashed boost folder
+	if os.path.exists(stash_boost_path + "/CMakeLists.txt"):
```
I would make a couple more changes for clarity. 
Combining os.path.abspath (or relpath) with os.path.join would make full path names a little more maintainable. Even though os.path treats paths as strings, we should strive to use its methods rather than depend on string literals that hard-code the "path-y" elements, like slashes.

spahrenk committed:
Improve naming when updating boost.

Enabled renaming of both stash and destination boost folders, if needed. Added function to handle naming conflicts.

spahrenk committed:
Make cosmetic changes.

spahrenk committed:
Modify update-boost.py to use os.path.join

spahrenk commented:
update-boost.py modified to use os.path.join for concatenation.

tanaya-mankad commented on scripts/update-boost.py:
```diff
@@ -80,44 +80,48 @@ def find_and_replace(filename, word, replacement):
 # repo_path: path to repo
 # prev_repo_boost_folder_name: name of current boost folder (within repo)
 # dest_repo_boost_folder_name: new name for updated, minimized boost folder (within repo)
 def updateBoost(bcp_command, source_boost_path, repo_path, prev_repo_boost_folder_name, dest_repo_boost_folder_name):
```
`update_boost `for self-consistent (and PEP8 compliant) naming

tanaya-mankad commented on scripts/update-boost.py:
```diff
@@ -129,36 +133,36 @@ def updateBoost(bcp_command, source_boost_path, repo_path, prev_repo_boost_folde
+		shutil.copyfile(os.path.join(stash_boost_path, "CMakeLists.txt"), os.path.join(dest_repo_boost_path, "CMakeLists.txt"))
 	else:
-		print("\"" + stash_boost_path + "/CMakeLists.txt\" does not exist")
+		print("\"" + os.path.join(stash_boost_path, "CMakeLists.txt") + "\" does not exist")
```
We should probably avoid repeating the literal `os.path.join(stash_boost_path, "CMakeLists.txt")` so many times.

tanaya-mankad commented on scripts/update-boost.py:
```diff
@@ -129,36 +133,36 @@ def updateBoost(bcp_command, source_boost_path, repo_path, prev_repo_boost_folde
-		if os.path.exists(dest_repo_boost_path + "/libs/program_options"):
-			print("Copying  \"" + stash_boost_path + "/libs/program_options/CMakeLists.txt\"")
-			shutil.copyfile(stash_boost_path + "/libs/program_options/CMakeLists.txt", dest_repo_boost_path + "/libs/program_options/CMakeLists.txt")
+	if os.path.exists(os.path.join(stash_boost_path, "libs/program_options/CMakeLists.txt")):
```
`os.path.join(stash_boost_path, "libs/program_options/CMakeLists.txt")` isn't taking full advantage of join. Better would be `os.path.join(stash_boost_path, "libs", "program_options", "CMakeLists.txt")`
(See also above comment about repeating.)

spahrenk committed:
Improve update-boost.py

Updated heading documentation. Improved uniformity of displayed messages. Reduced number of join calls.

spahrenk commented:
Calls to os.path.join were reorganized and reduced for clarity. Displayed status reports were changed to conform with the format used by bcp and for uniformity. Heading notes were altered for consistency and to incorporate latter changes.

tanaya-mankad reviewed (approved):
One last quick thing: (since it hadn't changed in a recent commit, I wonder if Github skipped putting my comment in the review...)
Line 95 `updateBoost `should probably be `update_boost `for self-consistency.

spahrenk committed:
Update update-boost.py

updateBoost->update_boost

spahrenk commented:
Renamed updateBoost->update_boost

spahrenk committed:
Update to boost-1.85.0.

spahrenk committed:
Revert to boost-1.83.0.

spahrenk committed:
Revert boost (2).

spahrenk committed:
Change to Ruby 3.0

spahrenk committed:
Change to File.exist.

nealkruis committed:
Update platform/build matrix.

nealkruis committed:
Revise platform/build matrix.

nealkruis committed:
Fix coverage build.

nealkruis committed:
Add weekly builds, modify coverage build, temporarily disable macos-13 build.
